cpp/src/parquet/file_reader.cc

```diff
@@ -383,6 +382,9 @@ void SerializedFile::ParseMetaDataOfEncryptedFileWithPlaintextFooter(
     file_decryptor_.reset(new InternalFileDecryptor(
```

Comment:
It would nice to push this inside `FileMetaData` to avoid this dependency cycle. `set_file_decryptor` can be removed.

Reply:
@majetideepak The thing is that in the case of an Encrypted file with a Plaintext footer, we first need to parse the `FileMetaData` thrift message to get the `footer_signing_key_metadata` - so we can’t create an instance of `InternalFileDecryptor` prior to creating an instance of `FileMetaData` (unless we want to also deserialize the `FileMetaData` thrift message outside of `parquet::FileMetaData`).
We could create the `InternalFileDecryptor` inside `parquet::FileMetaData`, but it’s also used by `SerializedFile`, so i think it's Ok to leave its creation inside `SerializedFile`. what do you think?

Reply:
It is nice to have the `FileMetaData` object be immutable once its created. 
How about invoking the `ParseMetaDataOfEncryptedFileWithPlaintextFooter` call inside  `ParseUnencryptedFileMetadata`  and avoid this `set_file_descryptor` call? We are simply moving the `if-else` block from `ParseMetaData` to `ParseUnencryptedFileMetadata`

Reply:
I agree.
The only thing is that in order for the `InternalFileDecryptor` to be `const` in `parquet::FileMetaData` (actually `FileMetaDataImpl`), we need to have the decryptor "ready" before the call to `parquet::FileMetaData::Make`, but specifically in the case of an encrypted file with a plaintext footer (from the [format](https://github.com/apache/parquet-format/blob/master/Encryption.md#55-plaintext-footer-mode)), we need to check if the encryption algorithm is set, and if so retrieve the `footer_signing_key_metadata`

```thrift
struct FileMetaData {
...
  /** 
   * Encryption algorithm. This field is set only in encrypted files
   * with plaintext footer. Files with encrypted footer store algorithm id
   * in FileCryptoMetaData structure.
   */
  8: optional EncryptionAlgorithm encryption_algorithm

  /** 
   * Retrieval metadata of key used for signing the footer. 
   * Used only in encrypted files with plaintext footer. 
   */ 
  9: optional binary footer_signing_key_metadata
}
```
To do that, we need to parse the `FileMetaData` thrift message - this is done when creating an instance of `parquet::FileMetaData`.
So, for creating a file decrpytor **before** the call to `parquet::FileMetaData::Make`, we need information that will only be available **after** the call.

If we want the decryptor to be const in `parquet::FileMetaData`, we can take a few approaches:
1) Partially parse the `FileMetaData` thrift message before the call to `parquet::FileMetaData::Make`, then create the decryptor and call `parquet::FileMetaData::Make` (which will parse it again. we can alternatively also reuse the parsed thrift message).
2) Create the decryptor inside `FileMetaData` (instead of getting it as argument) - this requires passing the `FileDecryptionProperties` to it (which is also less "clean" IMHO). the issue with this approach is that the file decryptor is used in `SerializedFile` (which creates the `parquet::FileMetaData`), so we'll need to create the `parquet:FileMetaData`, then retrieve the decryptor and set it also in `SerializedFile`, seems less "clean" to me.
3) Make `InternalFileDecryptor` mutable, so we can change the algorithm and `footer_signing_key_metadata` after it's created. also looks less clean and just moves the problem to another area.

each of these options will allow `parquetFileMetaData` to be immutable, but has its own tradeoff.

Reply:
Thanks for clarifying this. I see the dependency issue now. This choice is fine. However, I see a liveness/ownership problem. `FileMetaData` is a shared pointer inside `SerializedFile`. We need to make `InternalFileDecryptor` a `shared_ptr` inside `SerializedFile` as well and set it as a `shared_ptr` inside the metadata.

Reply:
Done.

Reply:
Done